SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialShadowActor.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+#include "EngineClasses/SpatialShadowActor.h"
+#include "EngineClasses/SpatialNetDriverAuthorityDebugger.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialShadowActor);
+
+void USpatialShadowActor::Init(AActor& InActor)
+{
+	NumReplicatedProperties = 0;
+
+	for (TFieldIterator<FProperty> PropIt(InActor.GetClass()); PropIt; ++PropIt)
+	{
+		FProperty* Property = *PropIt;
+
+		if (Property->HasAnyPropertyFlags(CPF_Net) && Property->HasAnyPropertyFlags(CPF_HasGetValueTypeHash))
+		{
+			NumReplicatedProperties++;
+		}
+	}
+
+	// Intialise the array
+	ReplicatedPropertyHashes.Reserve(NumReplicatedProperties);
+	for (int32 i = 0; i < NumReplicatedProperties; i++)
+	{
+		ReplicatedPropertyHashes.Add(0);
+	}
+
+	CreateHash(InActor);
+}
+
+void USpatialShadowActor::Update(AActor& InActor)
+{
+	CreateHash(InActor);
+}
+
+void USpatialShadowActor::CreateHash(const AActor& InActor)
+{
+	int32 i = 0;
+	// Store a hashed value for each replicated property
+	for (TFieldIterator<FProperty> PropIt(InActor.GetClass()); PropIt; ++PropIt)
+	{
+		FProperty* Property = *PropIt;
+
+		if (Property->HasAnyPropertyFlags(CPF_Net) && Property->HasAnyPropertyFlags(CPF_HasGetValueTypeHash))
+		{
+			ReplicatedPropertyHashes[i] = Property->GetValueTypeHash(Property->ContainerPtrToValuePtr<void>(&InActor, 0));
+			i++;
+		}
+	}
+}
+
+void USpatialShadowActor::CheckUnauthorisedDataChanges(const AActor& InActor)
+{
+	check(!InActor.HasAuthority());
+
+	if (USpatialNetDriverAuthorityDebugger::IsSuppressedActor(InActor))
+	{
+		// We are suppressing warnings about some actor classes to avoid spamming the user
+		return;
+	}
+
+	// Compare hashed properties
+	int32 i = 0;
```

Comment:
Whilst I don't *think* properties can appear or disappear at runtime, but I'd have to double check, propery flags might be able to change? Again I'd have to check, but if they can, then the indexing method used here could break.

It might be safer to store a TMap<FProperty*,uin32> perhaps?

Reply:
In theory you "might" be able to get away with just combining the hashes into a single value, if the order or number or types of properties changed then the single overall hash value would change... I don't think you'll get any clashes, hopefully...

Reply:
If you combine the hashes into a single property you cannot easily tell which property changed and would also break if the indexing changed as mentioned in the first point.